File: Week7/ProblemA.c
/*
Language: c
ID: 63386
QLink: https://quera.org/course/assignments/63386/problems
Author: @Amirkhaskar
*/

#include <stdio.h>
#define max(a, b) (((a) > (b)) ? (a) : (b))
#define MAX 1000

int maxProfit(int prices[], int length, int k) {
    int n = length;
    int profit[MAX][MAX];
    for (int i = 0; i <= k; i++) {
        profit[i][0] = 0;
    }
    for (int i = 0; i <= n; i++) {
        profit[0][i] = 0;
    }
    for (int i = 1; i <= k; i++) {
        for (int j = 1; j < n; j++) {
            int currMax = 0;
            for (int day = 0; day < j; day++) {
                currMax = max(currMax, prices[j] - prices[day] + profit[i - 1][day]);
            }
            profit[i][j] = max(profit[i][j - 1], currMax);
        }
    }
    return profit[k][n - 1];
}

int main() {
    int k;
    scanf("%d", &k);
    int num=0;
    int i=0;
    int x[1000];
    while (scanf("%d", &num) == 1) {
        x[i] = num; 
        i++; 
    }
    printf("%d", maxProfit( x,i,k ));
    return 0;
}

File: Week7/ProblemA.java
/*
Language: java
ID: 63386
QLink: https://quera.org/course/assignments/63386/problems
Author: @Amirkhaskar
*/
import java.util.Scanner;


public class Main
{
  public static int maxProfit (int k, int[]prices)
  {
    int len = prices.length;
    if (k >= len / 2)
        return quickSolve (prices);

    int[][] t = new int[k + 1][len];
    for (int i = 1; i <= k; i++)
      {
	int tmpMax = -prices[0];
	for (int j = 1; j < len; j++)
	  {
	    t[i][j] = Math.max (t[i][j - 1], prices[j] + tmpMax);
	    tmpMax = Math.max (tmpMax, t[i - 1][j - 1] - prices[j]);
	  }
      }
    return t[k][len - 1];
  }


  private static int quickSolve (int[]prices)
  {
    int len = prices.length, profit = 0;
    for (int i = 1; i < len; i++)
      // as long as there is a price gap, we gain a profit.
      if (prices[i] > prices[i - 1])
	profit += prices[i] - prices[i - 1];
    return profit;
  }

 
  public static void main (String[]args)
  {
        Scanner scanner = new Scanner(System.in);

        // Read k in one line
        int k = scanner.nextInt();
        scanner.nextLine(); // Move to the next line

        // Read array of prices as space-separated values
        String[] pricesStr = scanner.nextLine().split(" ");
        int[] prices = new int[pricesStr.length];

        // Convert string array to int array
        for (int i = 0; i < pricesStr.length; i++) {
            prices[i] = Integer.parseInt(pricesStr[i]);
        }

        // Calculate and print the maximum profit
        int maxProfit = maxProfit(k, prices);
        System.out.println(maxProfit);
  }


}
